src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/CrossChainTransferStore.cs

```diff
@@ -177,155 +204,234 @@ private HashHeightPair TipToChase()
+                    tracker.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
+                }
+            }
 
-                    CrossChainTransfer[] partialTransfers = this.Get(dbreezeTransaction,
-                        this.depositsIdsByStatus[CrossChainTransferStatus.Partial].Union(
-                            this.depositsIdsByStatus[CrossChainTransferStatus.FullySigned]).ToArray());
+            if (tracker.Count == 0)
+                return;
+
+            using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
+            {
+                dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
 
-                    foreach (CrossChainTransfer partialTransfer in partialTransfers)
+                try
+                {
+                    foreach (KeyValuePair<ICrossChainTransfer, CrossChainTransferStatus?> kv in tracker)
                     {
-                        if (!SanityCheck(partialTransfer.PartialTransaction, wallet))
-                        {
-                            this.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
-                            await this.PutTransferAsync(dbreezeTransaction, partialTransfer);
-                        }
+                        this.PutTransfer(dbreezeTransaction, kv.Key);
                     }
 
                     dbreezeTransaction.Commit();
+
+                    this.UpdateLookups(tracker);
+                }
+                catch (Exception err)
+                {
+                    this.logger.LogError("Error during database update: {0}", err.Message);
+                    this.logger.LogTrace("(-):[SANITY_ERROR]");
+
+                    // Restore expected store state in case the calling code retries / continues using the store.
+                    dbreezeTransaction.Rollback();
+                    throw err;
                 }
             }
+        }
 
-            this.logger.LogTrace("(-)");
+        public static Transaction BuildDeterministicTransaction(uint256 depositId, Recipient.Recipient recipient, Network network,
+            IFederationWalletTransactionHandler federationWalletTransactionHandler, string walletPassword = "", Transaction[] partials = null, ILogger logger = null)
+        {
+            try
+            {
+                logger?.LogTrace("()");
+
+                uint256 opReturnData = depositId;
+
+                // Build the multisig transaction template.
+                var multiSigContext = new Recipient.TransactionBuildContext(new[] { recipient }.ToList(), opReturnData: opReturnData.ToBytes())
+                {
+                    TransactionFee = Money.Coins(0.01m), // TODO: Configurable?
+                    MinConfirmations = 1,                // TODO: This may have to be zero to allow multiple deposits spending a single UTXO.
+                    Shuffle = false,
```

Comment:
Are these TODO's part of the cross-chain-transfer work ? Are they needed for this release ?

Reply:
Yes.

Reply:
Maybe we need a card on the board for this then - @monsieurleberre, @bokobza ?

Reply:
I've added the settings and implemented them. We can remove them if it turns out we need to.